[RISCV] Inline Assembly Support for GPR Pairs ('R')

lenary opened the pull request:
This patch adds support for getting even-odd general purpose register
pairs into and out of inline assembly using the `R` constraint as
proposed in riscv-non-isa/riscv-c-api-doc#92

There are a few different pieces to this patch, each of which need their
own explanation.

- Renames the Register Class used for f64 values on rv32i_zdinx from
  `GPRPair*` to `GPRF64Pair*`. These register classes are kept broadly
  unmodified, as their primary value type is used for type inference
  over selection patterns. This rename affects quite a lot of files.

- Adds new `GPRPair*` register classes which will be used for `r`
  constraints and for instructions that need an even-odd GPR pair. This
  new type is used for `amocas.d.*`(rv32) and `amocas.q.*`(rv64) in
  Zacas, instead of the `GPRF64Pair` class being used before.

- Marks the new `GPRPair` class legal as for holding a `MVT::Untyped`.
  Two new RISCVISD node types are added for creating and destructing a
  pair - `BuildGPRPair` and `SplitGPRPair`, and are introduced when
  bitcasting to/from the pair type and `untyped`.

- Adds functionality to `splitValueIntoRegisterParts` and
  `joinRegisterPartsIntoValue` to handle changing `i<2*xlen>` MVTs into
  `untyped` pairs.

- Adds an override for `getNumRegisters` to ensure that `i<2*xlen>`
  values, when going to/from inline assembly, only allocate one (pair)
  register (they would otherwise allocate two). This is due to a bug in
  SelectionDAGBuilder.cpp which other backends also work around.

- Ensures that Clang understands that `R` is a valid inline assembly
  constraint.

- This also allows `R` to be used for `f64` types on `rv32_zdinx`
  architectures, where doubles are stored in a GPR pair.

topperc commented:
Is the plan to make i128 a legal type for RV64?

lenary commented:
> Is the plan to make i128 a legal type for RV64?

I think that whichever type I choose for pairs, I will probably have to end up making legal? I'm still working through that. I know right now my testcases crash because the pair type is not legal, but I think I have the opportunity to break up the pair type into two registers in `RISCVTargetLowering::splitValueIntoRegisterParts` (or there are other hooks if i choose to represent it with `v2ixlen`). Having said that, I might just be pushing the problem deeper into ISel if I try that.

I remember I had an absolute nightmare with constraints for fp16/bf16 in the ARM target in 2022, probably again because of type legality, and I'm keen to avoid that this time.

topperc commented:
> > Is the plan to make i128 a legal type for RV64?
> 
> I think that whichever type I choose for pairs, I will probably have to end up making legal? I'm still working through that. I know right now my testcases crash because the pair type is not legal, but I think I have the opportunity to break up the pair type into two registers in `RISCVTargetLowering::splitValueIntoRegisterParts` (or there are other hooks if i choose to represent it with `v2ixlen`). Having said that, I might just be pushing the problem deeper into ISel if I try that.
> 
> I remember I had an absolute nightmare with constraints for fp16/bf16 in the ARM target in 2022, probably again because of type legality, and I'm keen to avoid that this time.

I'm concerned that if we make i128 legal, we will have to write custom splitting code for every operation. Using v2i64 probably has other problems conflicting with the V extension fixed vector support.

The Arch64 MVT::i64x8 may be a way to work around this. We could have a special pair MVT only used by inline assembly.

lenary commented:
> I'm concerned that if we make i128 legal, we will have to write custom splitting code for every operation. Using v2i64 probably has other problems conflicting with the V extension fixed vector support.

I sort-of expected that making `i128` legal would be problematic, I didn't realise that `v2i64` would also conflict with something we're currently using.

> The Arch64 MVT::i64x8 may be a way to work around this. We could have a special pair MVT only used by inline assembly.

Ok, this sounds doable - so something like `MVT::xlen_pair` is the way I'll lean, rather than a different type for rv32/rv64.

lenary commented:
> > The Arch64 MVT::i64x8 may be a way to work around this. We could have a special pair MVT only used by inline assembly.
> 
> Ok, this sounds doable - so something like `MVT::xlen_pair` is the way I'll lean, rather than a different type for rv32/rv64.

Turns out that MVTs really want a size (unless they're overloadable, which is a complex and under-documented property), so I'll be going with two different MVTs rather than trying to share one.

lenary commented:
So, I've just pushed an update:
- This now adds the two new MVTs, as suggested by Craig. I'm a little concerned if this causes a problem of using too much space in SelectionDAG tables. It's definitely a concern.
- I'm using them from `getAsmOperandValueType`, which is the right place to say "use this type for an inline asm operand" when it doesn't fit into the calling convention.
- I've added `BuildXLenPair` and `SplitXLenPair` ISD nodes, to represent building a pair and splitting it. Only the former is used so far. The names are bad, I'll work on better ones (`BuildPairGPR` etc is my current lean)
- The implementation in `RISCVTargetLowering::splitValueIntoRegisterParts` is, as far as I can tell, now correct. The SelectionDAG looks right for inline assembly inputs.
- I added a case to the switch in `RISCVDAGToDAGISel::Select` because that seemed easier than typing a lot of SelectionDAG stuff. I will probably turn this into TableGen tomorrow.


I don't know whether I got too tired, or something else, but the implementation of `RISCVTargetLowering::joinRegisterPartsIntoValue` has not been making sense for me at all. I'd expect this to have an almost identical check as in `splitValueIntoRegisterParts`, and effectively do the opposite, but it doesn't. I think both `PartVT` and `ValueVT` end up as `riscv_i64_pair` for the rv64 testcase, once the `_out` function is uncommented. One of these comes from the constraint's RegClass's first legal type (defensible), the other I'm a bit more confused by, and think it might be coming from `getAsmOperandValueType`, but I'm not sure I traced this back correctly. I'll look again tomorrow.

lenary commented:
I think the first thing I might do tomorrow is some of the NFC cleanups and commenting that is in this PR.

I'm not sure the register class renaming can happen quite yet.

lenary commented:
I gave up on using `splitValueIntoRegisterParts` and `joinRegisterPartsIntoValue`, and instead added a custom legalisation of `BITCAST` with the 2*xlen type, which matches what we do for `BuildPairF64` and `SplitF64`. This ended up being much more "symmetrical", which was nice.

I'm still doing Selection entirely in C++, as I think we'd need to go to C++ anyway to emit the right REG_SEQUENCE and EXTRACT_SUBREG.

I've not implemented anything in `RISCVTargetLowering::PerformDAGCombine` for `BuildPairGPR` and `SplitPairGPR` - though I see there's something in there for `SplitF64`, and not all of it is FP-related. I don't know how much of a difference these combines will make when the generated instructions are not as expensive as the stack slot that `BuildPairF64`/`SplitF64` sometimes needs.

Instead of looking at combines, I'm going to try to get the tests to work with GlobalISel.

lenary commented:
NVM about GlobalISel "Inline asm lowering is not supported for this target yet".

lenary committed:
[RISCV] GPR Pairs for Inline Asm using `Pr`

This patch adds support for getting even-odd general purpose register
pairs into and out of inline assembly using the `Pr` constraint as
proposed in riscv-non-isa/riscv-c-api-doc#92

There are a few different pieces to this patch, each of which need their
own explanation.

- Renames the Register Class used for f64 values on rv32i_zdinx from
  `GPRPair*` to `GPRF64Pair*`. These register classes are kept broadly
  unmodified, as their primary value type is used for type inference
  over selection patterns. This rename affects quite a lot of files.

- Adds new `GPRPair*` register classes which will be used for `Pr`
  constraints and for instructions that need an even-odd GPR pair. This
  new type is used for `amocas.d.*`(rv32) and `amocas.q.*`(rv64) in
  Zacas, instead of the `GPRF64Pair` class being used before.

- Marks the new `GPRPair` class legal as for holding a `MVT::Untyped`.
  Two new RISCVISD node types are added for creating and destructing a
  pair - `BuildGPRPair` and `SplitGPRPair`, and are introduced when
  bitcasting to/from the pair type and `untyped`.

- Adds functionality to `splitValueIntoRegisterParts` and
  `joinRegisterPartsIntoValue` to handle changing `i<2*xlen>` MVTs into
  `untyped` pairs.

- Adds an override for `getNumRegisters` to ensure that `i<2*xlen>`
  values, when going to/from inline assembly, only allocate one (pair)
  register (they would otherwise allocate two). This is due to a bug in
  SelectionDAGBuilder.cpp which other backends also work around.

- Ensures that Clang understands that `Pr` is a valid inline assembly
  constraint.

- Adds Conditions to the GPRF64Pair-related changes to `LowerOperation`
  and `ReplaceNodeResults` which match when BITCAST for the relevant
  types should be handled in a custom manner.

- This also allows `Pr` to be used for `f64` types on `rv32_zdinx`
  architectures, where doubles are stored in a GPR pair.

lenary commented:
Sorry, got the rebase slightly wrong. Will re-do it tomorrow. Tests will likely fail in the meantime. This only really affects the branch-relaxation tests.

lenary commented:
Today's changes address all the CI failures, by adding better checking before turning BITCAST into `BuildPairF64` or `SplitF64` - the conditions are derived from the conditions on `setOperationAction(ISD::BITCAST, MVT::i64, Custom)` before this patch added more of the same.

lenary commented:
ping?

wangpc-pp reviewed (commented):
Thanks! I just had a detailed look. Given that you have explained almost all the code detailedly, I think this PR looks great to me!
Just some overall comments:
1. I personally like your proposal of adding new constraints, but we still need the agreement between community members.
2. I saw all the comments above and I know reason why we choose to add new MVT types. My question is, maybe we can make it less target-specific? I don't think this is a RISC-V only problem.
3. We should be able to use `Pr` for Zacas now? So maybe we should add some tests for it.

lenary commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -2224,6 +2231,17 @@ bool RISCVTargetLowering::isExtractSubvectorCheap(EVT ResVT, EVT SrcVT,
 
+EVT RISCVTargetLowering::getAsmOperandValueType(const DataLayout &DL, Type *Ty,
+                                                bool AllowUnknown) const {
+  if (!Subtarget.is64Bit() && Ty->isIntegerTy(64))
```
This file tends to mostly use `Subtarget.is64Bit()` rather than the `isRV64`/`isRV32`, which is what I followed here. This might get a lot more simple if I move to trying `MVT::Other` though.

lenary reviewed (commented):
> Thanks! I just had a detailed look. Given that you have explained almost all the code detailedly, I think this PR looks great to me! Just some overall comments:
> 
> 1. I personally like your proposal of adding new constraints, but we still need the agreement between community members.

I think some of the agreement needs an implementation, maybe? I'll ping the c-api issue.

> 2. I saw all the comments above and I know reason why we choose to add new MVT types. My question is, maybe we can make it less target-specific? I don't think this is a RISC-V only problem.

I noticed, after-the-fact, that SystemZ uses `MVT::Untyped` (corrected) for pairs, I can prepare a fixup commit that does this (and removes the `riscv_*_pair` MVTs), if we think that's a better target-independent approach?

> 3. We should be able to use `Pr` for Zacas now? So maybe we should add some tests for it.

- The test functions which have an in-out `Pr` parameter (`test_Pr_wide_scalar_inout`) are derived from the issue about `amocas.q` in inline assembly on RV64. The C from that issue found some bugs in the implementation, which is why those tests exist. These don't reference any specific instructions just to make them have less complex dependencies - which is why the text of the inline assembly is just a comment to verify we're producing sensible output.
- The Zacas pair instructions continue to use `GPRPair` in their definitions, rather than the new `GPRF64Pair` register classes, which seemed the right way forwards. There are no patterns for the pair instructions at the moment, but I'm not sure I want to change the codegen of atomics, that's a whole other nightmare. Do you have specific tests in mind?

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -2238,6 +2256,17 @@ MVT RISCVTargetLowering::getRegisterTypeForCallingConv(LLVMContext &Context,
 }
 
+unsigned
+RISCVTargetLowering::getNumRegisters(LLVMContext &Context, EVT VT,
```
Why did we need this change but AArch64 didn't?

lenary replied:
This is to prevent an assert that is hit in `RegsForValue::AddInlineAsmOperands` without this code.

I think AArch64 is not hitting this because they didn't test very much - the `amocas.q` example has a lot of hard cases (multiple returns, matching input/output operands) which hit a lot of the hard cases in SelectionDAGBuilder. Note I had to fix one place where AArch64 was definitely wrong - the changes to call `getAsmOperandValueType` when there are multiple outputs from asm - AArch64 should have hit this case and didn't.

I've spent a few more hours this evening trying to trace down this assert, and I feel closer, but not quite there. The testcase to use to get where I've got to (with the assert) is `test_Pr_wide_scalar_inout`.

SelectionDAGBuilder.cpp's `getRegistersForValue` seems to be doing the right thing, always calling this with the `VT` and `RegisterVT` with the same values, as far as I can tell. It seems to create the `OpInfo.AssignedRegs` in such a way that the later call to `getNumRegisters` in `RegsForValue::AddInlineAsmOperands` will get the same value as it did when it was created.

Where this seems to go wrong is the matching inputs code in SelectionDAGBuilder.cpp - https://github.com/llvm/llvm-project/blob/8b55162e195783dd27e1c69fb4d97971ef76725b/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp#L10119-L10139 

This seems to have its own logic about creating the `RegsForValue MatchedRegs(...)` based on the surrounding DAG Context, and seems to avoid the info available in `OpInfo` which should have some relevancy, instead this code is directly pulling out similar (but not quite identical) information from the DAG. When `getNumRegisters` is eventually called in `MatchedRegs::AddInlineAsmOperands`, it uses `MVT::i128` rather than `MVT::riscv_i64_pair`, which means it thinks it needs two registers, not just one. I think this `MVT::i128` is coming from `InOperandVal.getValueType()` when MatchedRegs is created.

When this code is compared to `getRegistersForValue`, the two really don't seem to line up, logically - I would expect them both to be doing similar things, but they're not - how the `RegsForValue` object is created in `getRegistersForValue` is reasonably different to this logic. 

git-blame-ing this code, it seems to have been introduced for supporting i128 on SystemZ - presumably also for paired operands - and they do have a `getNumRegisters` override, which returns `1` - https://reviews.llvm.org/D100788 (refactors have hit a good number of lines around the SelectionDAGBuilder.cpp visitInlineAsm code I believe to be at fault, but nothing that's not NFC since that change).

I don't feel good about the matching inputs code in SelectionDAGBuilder - I don't understand it, and I don't think my addition to `getNumRegisters` should be necessary if SelectionDAGBuilder.cpp worked more closely to how `getRegistersForValue` works. The extra confusing thing here is that `getRegistersForValue` might have mutated `OpInfo` when there's a matched input - but returned `std::nullopt`, so maybe the logic for matched inputs needs to just match the tail end of `getRegistersForValue`. I'm not entirely sure.

You're a lot more familiar with SelectionDAG than I am, do you have advice for what I should be doing here? It does seem like more target-independent code needs to be fixed to get rid of this. Maybe that can come later?

lenary commented:
Gentle Ping. I'm looking for answers to two questions:

- [Should I] prepare a fixup commit that [uses `MVT::Untyped`] (and removes the `riscv_*_pair` MVTs), if we think that's a better target-independent approach?

- Any advice on whether I should be digging deeply into changing SelectionDAGBuilder.cpp (affecting all targets) so we don't need the `getNumRegisters` override? The options here IMO are a) live with the override, b) change SelectionDAGBuilder now, c) change SelectionDAGBuilder in a follow-up and remove the override once it is not needed.

topperc commented:
> Gentle Ping. I'm looking for answers to two questions:
> 
> * [Should I] prepare a fixup commit that [uses `MVT::Untyped`] (and removes the `riscv_*_pair` MVTs), if we think that's a better target-independent approach?

I guess so. I didn't know about the SystemZ change before. I do find it a little weird having 2 register classes with the same spill size and registers. I believe the tablegen generated code considers them subclasses of each other. X86 has some register classes like this too so I guess its not a big deal.

> * Any advice on whether I should be digging deeply into changing SelectionDAGBuilder.cpp (affecting all targets) so we don't need the `getNumRegisters` override? The options here IMO are a) live with the override, b) change SelectionDAGBuilder now, c) change SelectionDAGBuilder in a follow-up and remove the override once it is not needed.

I think we can live with the override since SystemZ is doing it too.

lenary commented:
I'm working on this again and tripping over the commit that changes the branch relaxation tests, so I'm planning to just push that to `main` (it's NFC) and rebase. Sorry for the PR noise this will generate. I'll keep the "attempt to use `Untyped`" commit separate when I re-push to make reviews easier.

lenary commented:
> > Gentle Ping. I'm looking for answers to two questions:
> > 
> > * [Should I] prepare a fixup commit that [uses `MVT::Untyped`] (and removes the `riscv_*_pair` MVTs), if we think that's a better target-independent approach?
> 
> I guess so. I didn't know about the SystemZ change before. 

I'll try it, and push a separate commit, so we can see what else is needed.

>
> I do find it a little weird having 2 register classes with the same spill size and registers. I believe the tablegen generated code considers them subclasses of each other. X86 has some register classes like this too so I guess its not a big deal.

I'm not very happy about the duplication either, but it's because of a limitation where fairly early on, selectiondag uses the first legal type in a class, rather than the one you might want.

> > * Any advice on whether I should be digging deeply into changing SelectionDAGBuilder.cpp (affecting all targets) so we don't need the `getNumRegisters` override? The options here IMO are a) live with the override, b) change SelectionDAGBuilder now, c) change SelectionDAGBuilder in a follow-up and remove the override once it is not needed.
> 
> I think we can live with the override since SystemZ is doing it too.

lenary commented:
The new approach using MVT::Untyped is uploaded. Inspect commit https://github.com/llvm/llvm-project/pull/112983/commits/126e4ee22278532e0cedc38846fced6b30009e2e to see what changed to go from the `riscv_i<xlen>_pair` approach to the one using `untyped`.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVRegisterInfo.td:
```diff
@@ -74,7 +74,10 @@ def sub_gpr_odd  : SubRegIndex<32, 32> {
 } // Namespace = "RISCV"
 
-// Integer registers
+//===----------------------------------------------------------------------===//
```
Please land these comment/style changes first.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -21351,6 +21372,17 @@ bool RISCVTargetLowering::splitValueIntoRegisterParts(
+    // Pairs in Inline Assembly
+    MVT XLenVT = Subtarget.getXLenVT();
+    SDValue Lo, Hi;
+    std::tie(Lo, Hi) = DAG.SplitScalar(Val, DL, XLenVT, XLenVT);
```
```suggestion
    auto [Lo, Hi] = DAG.SplitScalar(Val, DL, XLenVT, XLenVT);
```

lenary replied:
Done.

wangpc-pp reviewed (approved):
Cheers! It seems using `untyped` saves some lines as well!
LGTM with nits.

topperc commented on llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:
```diff
@@ -952,14 +952,43 @@ void RISCVDAGToDAGISel::Select(SDNode *Node) {
+    ReplaceNode(Node, N);
+    return;
+  }
+  case RISCVISD::SplitGPRPair: {
```
Can we share this code with SplitF64?

lenary replied:
Sure. Happy to in a follow up.

lenary replied:
I ended up doing this as I have to do a big rebase to sort out the comments/reorg of RISCVRegisterinfo.td

topperc reviewed (approved):
LGTM we can do the code sharing as a follow up if we want

lenary commented:
I will rewrite the description of the PR momentarily, once I've rebased over the cleanup/comments to RISCVRegisterinfo.td

lenary commented:
The commit I just pushed up should be the final version, and not fundamentally different from what was reviewed (but with various NFC patches split out).

I think I do have a route towards merging GPRPair and GPRF64Pair, but I'll do that in a follow-up, as there's a mystery crash somewhere that I haven't tracked down yet.

lenary committed:
fixup! [RISCV] GPR Pairs for Inline Asm using `Pr`

lenary committed:
amend! [RISCV] GPR Pairs for Inline Asm using `Pr`

[RISCV] GPR Pairs for Inline Asm using `R`

This patch adds support for getting even-odd general purpose register
pairs into and out of inline assembly using the `R` constraint as
proposed in riscv-non-isa/riscv-c-api-doc#92

There are a few different pieces to this patch, each of which need their
own explanation.

- Renames the Register Class used for f64 values on rv32i_zdinx from
  `GPRPair*` to `GPRF64Pair*`. These register classes are kept broadly
  unmodified, as their primary value type is used for type inference
  over selection patterns. This rename affects quite a lot of files.

- Adds new `GPRPair*` register classes which will be used for `R`
  constraints and for instructions that need an even-odd GPR pair. This
  new type is used for `amocas.d.*`(rv32) and `amocas.q.*`(rv64) in
  Zacas, instead of the `GPRF64Pair` class being used before.

- Marks the new `GPRPair` class legal as for holding a `MVT::Untyped`.
  Two new RISCVISD node types are added for creating and destructing a
  pair - `BuildGPRPair` and `SplitGPRPair`, and are introduced when
  bitcasting to/from the pair type and `untyped`.

- Adds functionality to `splitValueIntoRegisterParts` and
  `joinRegisterPartsIntoValue` to handle changing `i<2*xlen>` MVTs into
  `untyped` pairs.

- Adds an override for `getNumRegisters` to ensure that `i<2*xlen>`
  values, when going to/from inline assembly, only allocate one (pair)
  register (they would otherwise allocate two). This is due to a bug in
  SelectionDAGBuilder.cpp which other backends also work around.

- Ensures that Clang understands that `R` is a valid inline assembly
  constraint.

- Adds Conditions to the GPRF64Pair-related changes to `LowerOperation`
  and `ReplaceNodeResults` which match when BITCAST for the relevant
  types should be handled in a custom manner.

- This also allows `R` to be used for `f64` types on `rv32_zdinx`
  architectures, where doubles are stored in a GPR pair.

lenary commented:
I won't be merging this today. Instead, two changes/updates:
- Moving to use `R` instead of `Pr`, because kito pointed out that GCC cannot use `P` for anything except immediate values (done as a fixup for the moment, so we can back it out if we're struggling to agree on letters)
- Removing a change that's being done in #116159

lenary commented:
Kito and I have agreed `R` is a good way forwards for GPR pairs, and the c-api-doc PR is approved (but not landed). I'll reword the description/message to update our constraint choice when I land this, which I hope to get to later today.
